content/posts/cjd-casc.md

```diff
@@ -0,0 +1,548 @@
+date: 2019-07-01T17:00:00-04:00
+showDate: true
+draft: true
+tags: ["jenkins","cloudbees jenkins distribution","gitops","kubernetes","kaniko","docker","cert-manager"]
+---
+
+In this blog post, we'll walk through creating a self-updating instance of [CloudBees Jenkins Distribution](https://www.cloudbees.com/products/cloudbees-jenkins-distribution) fully configured-as-code stored in a Git repository.
+
+We'll deploy the CJD master as a StatefulSet in a Kubernetes cluster, configure the master using the [Jenkins Configuration as Code plugin](https://github.com/jenkinsci/configuration-as-code-plugin), and set up a TLS certificate through [cert-manager](https://github.com/jetstack/cert-manager). Finally, we'll seed a Pipeline job that updates the master upon commit to the [Git repository](https://github.com/cb-technologists/cjd-casc) that contains the configuration - enabling GitOps for Jenkins itself.
+
+## Deploying CloudBees Jenkins Distribution in Kubernetes
+First, we'll need to deploy a Jenkins instance into a Kubernetes cluster. In this case, we'll use [Google Kubernetes Engine](https://cloud.google.com/kubernetes-engine/) to deploy a containerized version of CJD. To provision a cluster, we'll follow the Google Cloud documentation [here](https://cloud.google.com/kubernetes-engine/docs/how-to/creating-a-cluster). (**Note:** this blog post assumes prior installation of and basic familiarity with using `kubectl` to interact with a Kubernetes cluster.)
+
+Once the cluster has been provisioned and `kubectl` has been configured, we'll create a dedicated namespace for our CJD resources and update our `kubectl config` to use it by default:
+
+```bash
+kubectl create namespace cjd
+kubectl config set-context $(kubectl config current-context) --namespace cjd
+```
+
+We'll also need to ensure an ingress controller is deployed within the cluster. For this post, we'll assume the use of the [NGINX ingress controller](https://kubernetes.github.io/ingress-nginx/). Following the [Installation Guide](https://github.com/kubernetes/ingress-nginx/blob/master/docs/deploy/index.md), we'll manually deploy using a few `kubectl` commands:
+
+```bash
+# grant cluster-admin to user
+kubectl create clusterrolebinding cluster-admin-binding \ --clusterrole cluster-admin \ --user $(gcloud config get-value account)
+# deploy nginx ingress controller resources
+kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/static/mandatory.yaml
+kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/static/provider/cloud-generic.yaml
+```
+
+Next, let's look at the manifest file that will deploy the necessary resources for CJD using the [cjd.yaml](https://github.com/cb-technologists/cjd-casc/blob/master/cjd.yaml) manifest file.
+
+First, we create a ServiceAccount, a Role with the necessary permissions to manage agents and perform the required update actions, and a RoleBinding to connect the two.
```

Comment:
Perhaps format K8s objects as code - ie: "First, we create a `ServiceAccount`, a `Role`"